File: 0402-remove-k-digits/0402-remove-k-digits.cpp
class Solution {
public:
    string removeKdigits(string num, int k) {
        string result;
        for (int i = 0; i < num.size(); i++) {
            while (!result.empty() && k && result.back() > num[i]) {
                k--;
                result.pop_back();
            }
            if (num[i] == '0') {
                if (!result.empty()) {
                    result.push_back(num[i]);
                }
            } else {
                result.push_back(num[i]);
            }
        }
        while(k and !result.empty()) {
            k--;
            result.pop_back();
        }
        return result.empty() ? "0" : result;
    }
};

File: 0402-remove-k-digits/0402-remove-k-digits.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        stack = []
        for digit in num:
            while stack and k and stack[-1] > digit:
                stack.pop()
                k -= 1
            if digit == '0':
                if stack:
                    stack.append(digit)
            else:
                stack.append(digit)
        while k and stack:
            k -= 1
            stack.pop()
        return "".join(stack) if stack else "0"
        